<div style="border:solid green 3px; padding: 20px">

<b>Привет!</b>

Меня зовут Сороколетов Илья и я буду проверять твой проект. 

Предлагаю общаться на «**ты**», но если тебе будет комфортнее общаться на «**вы**», то сообщи об этом в комментариях)


Для твоего удобства, я буду выделять свои комментарии следующим образом:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера v1:</b> Самые важные замечания. Они указывают на ключевые моменты, которые влияют на конечный результат проекта. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v1:</b> Советы или замечания, которые помогут сделать твою работу лучше, но необязательны к выполнению.

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> Так я выделяю все остальные комментарии.</div>


Пометками <b>v1-v2-v3-...</b> я буду отмечать версию проверки. Так ты сможешь быстро найти мои новые комментарии.

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения, например вот так:
    
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
</div>

<b>Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии.</b> Если ты оставляешь свои комментарии, то делай это под моими, чтобы сообщения были расположены в хронологическом порядке. Всё это поможет выполнить повторную проверку твоего проекта быстрее. 


Перед отправкой работы я рекомендую нажимать Kernel -> Restart & Run All. Это перезапустит ядро и по очереди выполнит все ячейки. Так ты сможешь проверить, что всё работает корректно. Кнопка Kernel находится в панели сверху

Обязательно задавай вопросы если они возникнут, а я перехожу к проверке)

P.S. На всякий случай, я оставлю пустой шаблон для твоих комментариев ниже. Кликни два раза на мой комментарий, скопируй последние четыре строчки кода и вставляй их в пустую ячейку там, где ты хочешь оставить комментарий. Не забудь только перед этим сменить тип ячейки на Markdown. Быстро это можно сделать так: кликнуть на ячейку - нажать ESC - нажать M. 
    
    
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Место для твоего комментария
</div>

# Исследование надежности заемщиков.

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v1:</b> 
<br>Лучше оставляй перед проектом краткое описание задачи. Это поможет тебе быстро вспомнить о чем проект, если спустя время ты решишь к нему вернуться)
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Исправлено
</div>

### Шаг 1. Обзор данных

In [1]:
import pandas as pd
import numpy as np
import seaborn as sea
from nltk.stem import SnowballStemmer 
russian_stemmer = SnowballStemmer('russian') 

In [2]:
df = pd.read_csv('/datasets/data.csv') #Запишем данные из csv файла в датафрейм
df.info() #Выведем на экран информацию о датафрейме

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v1:</b> 
<br>Все библиотеки лучше импортировать отдельно в первой ячейке. Это позволит тебе в любой момент добавить ещё одну библиотеку без перезаписи переменных.
</div>
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Исправлено
</div>

Таблица состоит из `12` столбцов. Названия стобцов указаны корректно.

Число строк отличается, в `total_income` и `days_employed`. Что свидетельствует о том, что там есть пропущенные значения

Изучим наполнение.

In [3]:
display(df.head(10)) #Выведем первые 10 строк таблицы
display(df.tail(10)) #Выведем последние 10 строк таблицы

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

Мы видим столбы со следующими данными:
1. `children` - количество детей у заемщика
2. `days_employed` - трудовой стаж в днях
3. `dob_years` - возраст заемщика
4. `education` - образование заемщика
5. `educatin_id` - цифровая категория образования заемщика
6. `family_status` - семейное положение
7. `family_status_id` - цифровая категория семейного положения
8. `gender` - пол заемщика
9. `income_type` - источник заработка
10. `debt` - имеет ли заемщик задолженность по возврату кредита
11. `total_income` - ежемесячный доход заемщика
12. `purpose` - цель получение кредита

Сразу можно увидеть, что:
1. Трудовой стаж не может быть отрицательным
2. Трудовой стаж не может быть равен `945` лет как у заемщика с индексом `4`
3. Формат записи образования разный, что может вызывать проблемы при обработке
Начнем предобработку данных

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Отлично!
<br>Во время первичного анализа данных, может помочь метод describe(). Он покажет разброс значений, максимум, минимум, среднее и медиану. 
</div>

### Шаг 2.1 Заполнение пропусков

In [4]:
df.isna().sum() #Выведем на экран число пропущенных значений по столбцам

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Есть прощуенные значения, что подтверждает выводы сделанные по информации о таблице. 

В `total_income` и `days_employed` есть пропущенные значения, это ежемесячный доход и трудовой стаж в днях. Оба значения являются количественным, а значит, для того, чтобы они не влияли на результаты их необходимо заменить на медианные значения, которые не повредят дальнейшему анализу.


In [5]:
total_income_mean = df['total_income'].mean() #Вычисляем среднее в столбце total_income
df['total_income'] = df['total_income'].fillna(value=total_income_mean) #Заполним пропущенные значения в total_income
days_employed_mean = df['days_employed'].mean() #Вычисляем среднее в столбце days_employed
df['days_employed'] = df['days_employed'].fillna(value=days_employed_mean) #Заполним пропущенные значения в days_employed
print(df.isna().sum()) #Выведем на экран число пропущенных значений по столбцам 

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


Проверка показала, что мы успешно заменили пропущенные значения


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Всё верно! Также здесь можно заполнить пропуски более оптимальным способом - медианой, в зависимости от типа занятости. Для этого отлично подойдёт функция transform:

    df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform("median"))

<br>Она создаёт новый столбец, который содержит столько же строк, сколько и изначальный, но только вместо оригинальных значений в каждой строке содержится медиана посчитанная по заданной группе. После этого вместо пропусков в оригинальном столбце мы подставляем значения из созданного столбца, то есть медианы в зависимости от группы.
</div>

### Шаг 2.2 Проверка данных на аномалии и исправления.

In [6]:
df.describe() #Выведем на экран основную информацию о ко каждому стобцу датафрейма

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  133526.013419     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2518.168900     33.000000      1.000000   
50%        0.000000    -982.531720     42.000000      1.000000   
75%        1.000000   63046.497661     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  9.763293e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.564003e+05  
75%            1.000000      0.000000  1.955436e+05  
max            4.000000      1.000000  2.265604e+06

Мы можем увидеть следующие значение, которые требуют нашего внимания.
1. `children` - максимум `20` детей, минимум `-1`. И если `20` детей требуют проверки, например может быть там только `1` заемщик такой, что может быть правдой. То `-1` ребенок быть не может, скорее всего ошибка в знаке. `20` детей скорее всего обсусловлено тем, что сотрудник совершил опечатку и вместо `2.0` ввел `20`. Следовательно примем за правило, что если больше `10` детей, мы выделяем целую часть от деления на `10`. 
2. `days_employed` - стоит обратить внимание на максимальное значение и минимальное значение. `401755` дней это `1 100` лет трудового стажа, что определенно является ошибкой. Если посмотреть на значение процетиля `75%`, мы можем понять, что как минимум `25%` стажа внесено некорректно.
3. `dob_years` - максимум и среднее, имеет адекватные значение, но мы видим, что есть значение `0`. Заемщиком не может быть лицо моложе `18` лет, значит это ошибка.

Для начала займемся столбцом `children`

In [7]:
df.loc[df['children'] < 0, 'children'] *= -1 #Определим все отрицательные значения в столбце и сделаем их положительными
df.loc[df['children'] > 9, 'children'] = df['children'] // 10 #Выделяем целую часть от деления на 10 когда детей 10 и более
display(df.loc[df['children'] > 9]) #Выведем число детей больше 9 на экран для проверки

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b>
<br>Отличное решение!
</div>


Далее рассмотрим `days_employed`, нас привлекает выброс в данных и отрицательные значения. Если исправить отрицательные значения просто, то в случае с выбросом желательно понять, чем он вызван.

Судя по стажу в днях и его виду, можно говорить о том, что стаж в днях был расчитан системой, однако тогда число дней должно было быть равным. Мы выделяем теорию, что исходя из значения процетиля `75%` в `63046.5` мы можем утверждать, что стаж был переведен часы 

Судя по данным, мы можем увидеть, что будет несколько значений где стаж с превышает возраст. 
За критерий "ошибочности" значения следующее выражение `Стаж в годах > (Возраст - 18)`. Значения которые попадают в этот критерий мы удалим.

In [8]:
df.loc[df['days_employed'] < 0, 'days_employed'] *= -1 #Превращаем отрицательные значения в положительные
df['days_employed'] = df['days_employed'] / (365*24) #Приняв теорию, что стаж указан в часах, мы вычисляем года стажа
df.loc[df['days_employed'] > (df['dob_years'] - 18), 'days_employed'] = np.nan #Удалим ошибочные значения
days_employed_mean = df['days_employed'].mean() #Вычислим среднее по столбцу
df['days_employed'] = df['days_employed'].fillna(value=days_employed_mean) #Заменим удаленные значения средним

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>👏🏻👏🏻👏🏻 Отличная работа с аномальными значениями. Молодец!
</div>

In [9]:
print(df.isna().sum()) #Проверим все ли пропущенные значения заменились

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


Далее займемся `dob_years`. Выясним, сколько людей меньше `18` лет.

In [10]:
display(df.loc[df['dob_years'] < 18].head(50)) #Выводим первые 50 значений где возраст меньше 18
display(df.loc[df['dob_years'] < 18].tail(50)) #Выводим последние 50 значений где возраст меньше 18

children  days_employed  dob_years            education  education_id  \
99            0       4.557863          0              Среднее             1   
149           0       4.557863          0              среднее             1   
270           3       4.557863          0              среднее             1   
578           0       4.557863          0              среднее             1   
1040          0       4.557863          0               высшее             0   
1149          0       4.557863          0              среднее             1   
1175          0       4.557863          0              среднее             1   
1386          0       4.557863          0               высшее             0   
1890          0       4.557863          0               высшее             0   
1898          0       4.557863          0              среднее             1   
2082          0       4.557863          0              среднее             1   
2284          0       4.557863          0              среднее             1   
2469          0       4.557863          0              среднее             1   
2487          0       4.557863          0              среднее             1   
2870          0       4.557863          0              среднее             1   
2941          2       4.557863          0              среднее             1   
3044          2       4.557863          0               высшее             0   
3147          2       4.557863          0               высшее             0   
3704          0       4.557863          0              среднее             1   
4064          1       4.557863          0              среднее             1   
4147          0       4.557863          0              среднее             1   
4922          0       4.557863          0               высшее             0   
4930          2       4.557863          0               высшее             0   
5014          0       4.557863          0              среднее             1   
6071          1       4.557863          0              среднее             1   
6407          2       4.557863          0              среднее             1   
6411          0       4.557863          0               высшее             0   
6670          0       4.557863          0               Высшее             0   
6778          0       4.557863          0               высшее             0   
6831          1       4.557863          0               высшее             0   
6848          2       4.557863          0              среднее             1   
6859          1       4.557863          0  неоконченное высшее             2   
7034          0       4.557863          0               высшее             0   
7073          0       4.557863          0              среднее             1   
7075          1       4.557863          0               высшее             0   
7252          0       4.557863          0               высшее             0   
7344          0       4.557863          0              среднее             1   
8061          0       4.557863          0               высшее             0   
8416          2       4.557863          0              среднее             1   
8574          0       4.557863          0              среднее             1   
8672          0       4.557863          0              среднее             1   
8784          0       4.557863          0              среднее             1   
9062          1       4.557863          0              среднее             1   
9151          0       4.557863          0               высшее             0   
9588          0       4.557863          0               высшее             0   
10163         0       4.557863          0              Среднее             1   
10188         0       4.557863          0              среднее             1   
10306         0       4.557863          0               высшее             0   
10364         2       4.557863          0               высшее             0   
10384  

children  days_employed  dob_years            education  education_id  \
10595         1       4.557863          0              среднее             1   
11077         0       4.557863          0              среднее             1   
11289         1       4.557863          0  неоконченное высшее             2   
11468         1       4.557863          0               высшее             0   
11481         2       4.557863          0              среднее             1   
11576         0       4.557863          0               высшее             0   
11664         0       4.557863          0              среднее             1   
11990         1       4.557863          0              СРЕДНЕЕ             1   
12062         0       4.557863          0              среднее             1   
12225         1       4.557863          0              среднее             1   
12297         0       4.557863          0              среднее             1   
12403         3       4.557863          0              среднее             1   
12729         0       4.557863          0              среднее             1   
13117         0       4.557863          0               высшее             0   
13187         0       4.557863          0              СРЕДНЕЕ             1   
13232         0       4.557863          0               высшее             0   
13413         0       4.557863          0               ВЫСШЕЕ             0   
13423         1       4.557863          0               высшее             0   
13439         0       4.557863          0              среднее             1   
13521         0       4.557863          0               высшее             0   
13741         0       4.557863          0              среднее             1   
13968         1       4.557863          0              среднее             1   
14477         0       4.557863          0              среднее             1   
14514         0       4.557863          0              среднее             1   
14608         0       4.557863          0              среднее             1   
14659         0       4.557863          0              среднее             1   
15140         2       4.557863          0              среднее             1   
15295         0       4.557863          0              среднее             1   
15409         2       4.557863          0              среднее             1   
15433         0       4.557863          0              среднее             1   
15886         0       4.557863          0              среднее             1   
15891         0       4.557863          0              среднее             1   
16042         2       4.557863          0              среднее             1   
16861         0       4.557863          0              среднее             1   
16922         0       4.557863          0               высшее             0   
17215         1       4.557863          0              среднее             1   
17306         0       4.557863          0              среднее             1   
17613         0       4.557863          0               высшее             0   
18176         0       4.557863          0              Среднее             1   
18415         0       4.557863          0              среднее             1   
18539         0       4.557863          0               Высшее             0   
18732         2       4.557863          0               высшее             0   
18851         0       4.557863          0              среднее             1   
19116         0       4.557863          0               высшее             0   
19371         1       4.557863          0               высшее             0   
19829         0       4.557863          0              среднее             1   
20462         0       4.557863          0              среднее             1   
20577         0       4.557863          0              среднее             1   
21179         2       4.557863          0               высшее             0   
21313  

Просмотр таблицы показал, что проблема с неверным значением стажем, коррелируется с нулем в возрасте, так как в данном случае мы видим значение среднего по стобцу стажа. 

Примем ноль в возрасте как отсутствующее значение и заменим все нули на средне значение возраста.

In [11]:
df.loc[df['dob_years'] == 0, 'dob_years'] = np.nan #Удалим все значения меньше 18
print('Число людей меньше 18 =', df['dob_years'].isna().sum())
dob_years_mean = df['dob_years'].mean() #Вычислим среднее по столцу
df['dob_years'] = df['dob_years'].fillna(value=dob_years_mean)

Число людей меньше 18 = 101


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v1:</b> 
<br>Не стоит делать замену ради замены, очень важна аргументация, почему ты подставляешь именно такое значение и насколько оно реалистично. Если мы заполняем пропуски таким образом, то у нас могут появится студенты которым за 40, да это возможно, но маловероятно. Так как нулевых значений в этом столбце меньше 1%, то их можно либо удалить либо вообще не трогать, так как сам столбец нам не нужен для дальнейшего исследования
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Я решил не удалять данные значения, чтобы оставить возможность расширения исследования, например если будет принято решение исследовать зависимость вида дохода и просрочек.
Так как таких значений 101, мы можем сказать, что их изменение с большой долей вероятности не исказит данные по просрочкам, в случае необходимости искать зависимость по возрасту, однако они могут нам пригодиться в исследовании вида дохода и просрочек.
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b>
<br>Хорошее решение!
</div>

Проверим успешность замены

In [12]:
df.isna().sum() #Проверим успешность замены

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Значения заменены

### Шаг 2.3. Изменение типов данных.

Переведем значения дохода в целочисленный тип, так как при 

In [13]:
df['total_income'] = df['total_income'].astype(int) #Переведем доход в целочисленный тип
display(df)

children  days_employed  dob_years education  education_id  \
0             1       0.963205       42.0    высшее             0   
1             1       0.459452       36.0   среднее             1   
2             0       0.641943       33.0   Среднее             1   
3             3       0.470862       32.0   среднее             1   
4             0       4.557863       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1       0.517045       43.0   среднее             1   
21521         0      39.262261       67.0   среднее             1   
21522         1       0.241250       38.0   среднее             1   
21523         3       0.355306       38.0   среднее             1   
21524         2       0.226542       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21520        224791             операции с жильем  
21521        155999          сделка с автомобилем  
21522         89672                  недвижимость  
21523        244093  на покупку своего автомобиля  
21524         82047         на покупку автомобиля  

[21525 rows x 12 columns]

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Тут всё верно!
</div>

### Шаг 2.4. Удаление дубликатов.

Начнем с приведения значений в столбцах к общему виду. Ознакомимся с тем, как написаны значения:

In [14]:
display(df['education'].unique()) #Выведем на экран уникальные значения в столбце
display(df['family_status'].unique()) #Выведем на экран уникальные значения в столбце
display(df['income_type'].unique()) #Выведем на экран уникальные значения в столбце
display(df['purpose'].unique()) #Выведем на экран уникальные значения в столбце

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Как мы видим, неявные дубликаты есть только в стобце `education`. Избавимся от них: 

In [15]:
df['education'] = df['education'].str.lower() #Приведем все значения в столбце к нижнему регистру
display(df['education'].unique()) #Выведем все уникальные значения в стобце

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Мы привели значения категоральные значения в столбцах к единому виду, и тем самым избавились от неявных дубликатов по значениям столбца. Теперь уберем явные дубликаты строк

In [16]:
print(len(df)) #Выведем длину датафрейма до удаления
df = df.drop_duplicates().reset_index(drop=True) #Удалим все дубликаты строк и обновим индексацию
print(len(df)) #Выведем длину датафрейма после удаления

21525
21454


Был удален `71` дубликат.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Супер! От дубликатов тоже избавились
</div>

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Для удобства просмотра таблицы сформируем поля `education` и `family_status` в отдельные словари, так как в таблице им уже присвоено id, а значит их можно выделить в отдельные словари используя эти же id.

In [17]:
edu_dict = df[['education','education_id']] #Сформируем отдельный датафрейм из двух столбцов
edu_dict = edu_dict.drop_duplicates().reset_index(drop=True) #Удалим все дубликаты и сбросим значения индекса
df.drop('education', inplace=True, axis=1) #Удалим столбец из изначального датафрейма
display(edu_dict) #Отобразим новый словарь
family_dict = df[['family_status','family_status_id']] #Сформируем отдельный датафрейм из двух столбцов
family_dict = family_dict.drop_duplicates().reset_index(drop=True) #Удалим все дубликаты и сбросим значения индекса
df.drop('family_status', inplace=True, axis=1) #Удалим столбец из изначального датафрейма
display(family_dict) #Отобразим новый словарь
display(df) #Проверим корректность удаления столбцов

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

children  days_employed  dob_years  education_id  family_status_id  \
0             1       0.963205       42.0             0                 0   
1             1       0.459452       36.0             1                 0   
2             0       0.641943       33.0             1                 0   
3             3       0.470862       32.0             1                 0   
4             0       4.557863       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21449         1       0.517045       43.0             1                 1   
21450         0      39.262261       67.0             1                 0   
21451         1       0.241250       38.0             1                 1   
21452         3       0.355306       38.0             1                 0   
21453         2       0.226542       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21449      F   компаньон     0        224791             операции с жильем  
21450      F   пенсионер     0        155999          сделка с автомобилем  
21451      M   сотрудник     1         89672                  недвижимость  
21452      M   сотрудник     1        244093  на покупку своего автомобиля  
21453      F   сотрудник     0         82047         на покупку автомобиля  

[21454 rows x 10 columns]

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Тут всё правильно 👌
</div>

### Шаг 2.6. Категоризация дохода.

На текущий момент столбец `total_income` не может нам сказать, как уровень дохода влияет на риск выдачи кредита, так как нет определния уровня дохода. В соотвествии с заданием определим категории дохода клиента.

`0–30000` — категория 'E'

`30001–50000` — категория 'D'

`50001–200000` — категория 'C'

`200001–1000000` — категория 'B'

`1000001 и выше` — категория 'A'.

In [18]:
def total_income_category(income): #Зададим функцию, которая проверяет значение на входе и возвращает категорию по данному значению 
    if income <= 30000:
        cat = 'E'
        return cat
    elif income <= 50000:
        cat = 'D'
        return cat
    elif income <= 200000:
        cat = 'C'
        return cat
    elif income <= 1000000:
        cat = 'B'
        return cat
    else:
        cat = 'A'
        return cat
df['total_income_category'] = df['total_income'].apply(total_income_category) #С помощью ф-ции apply применим функцию на каждую ячейку столбца total_income
display(df) #Проверим результат выполнения ф-ции

children  days_employed  dob_years  education_id  family_status_id  \
0             1       0.963205       42.0             0                 0   
1             1       0.459452       36.0             1                 0   
2             0       0.641943       33.0             1                 0   
3             3       0.470862       32.0             1                 0   
4             0       4.557863       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21449         1       0.517045       43.0             1                 1   
21450         0      39.262261       67.0             1                 0   
21451         1       0.241250       38.0             1                 1   
21452         3       0.355306       38.0             1                 0   
21453         2       0.226542       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21449      F   компаньон     0        224791             операции с жильем   
21450      F   пенсионер     0        155999          сделка с автомобилем   
21451      M   сотрудник     1         89672                  недвижимость   
21452      M   сотрудник     1        244093  на покупку своего автомобиля   
21453      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21449                     B  
21450                     C  
21451                     C  
21452                     B  
21453                     C  

[21454 rows x 11 columns]

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Категоризация клиентов по уровню дохода произведена верно
</div>

### Шаг 2.7. Категоризация целей кредита.

На текущий момент столбец `purpose` не может нам сказать, как цель кредита влияет на риск выдачи кредита, так как нет определения уровня дохода. В соотвествии с заданием определим категории цели взятия кредита как:

Операции с автомобилем

Операции с недвижимостью

Проведение свадьбы

Получение образования


In [19]:
def purpose_category(purpose): #Зададим функцию, которая проверяет значение на входе и возвращает категорию по данному значению 
    stemmed_query = russian_stemmer.stem(purpose) #Воспользуемся стеммингом для удобного поиска по словам
    for word in stemmed_query.split(' '): #Так как цель может состоять из несколько слов, нам нужно разделить строку на отдельные слова и проверить каждое.
        if word == 'автомобил' or word == 'автомоб':
            cat = 'операции с автомобилем'
            return cat
        elif word == 'недвижим' or word == 'жил' or word == 'жилья':
            cat = 'операции с недвижимостью'
            return cat
        elif word == 'свадьб':
            cat = 'проведение свадьбы'
            return cat
        elif word == 'образован':
            cat = 'получение образования'
            return cat
df['purpose_category'] = df['purpose'].apply(purpose_category) #С помощью ф-ции apply применим функцию на каждую ячейку столбца total_income
display(df[df['purpose_category'].isnull()]) #Проверим результат выполнения ф-ции, нет ли значений None  
display(df) #Проверим результат выполнения ф-ции      

Empty DataFrame
Columns: [children, days_employed, dob_years, education_id, family_status_id, gender, income_type, debt, total_income, purpose, total_income_category, purpose_category]
Index: []

children  days_employed  dob_years  education_id  family_status_id  \
0             1       0.963205       42.0             0                 0   
1             1       0.459452       36.0             1                 0   
2             0       0.641943       33.0             1                 0   
3             3       0.470862       32.0             1                 0   
4             0       4.557863       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21449         1       0.517045       43.0             1                 1   
21450         0      39.262261       67.0             1                 0   
21451         1       0.241250       38.0             1                 1   
21452         3       0.355306       38.0             1                 0   
21453         2       0.226542       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21449      F   компаньон     0        224791             операции с жильем   
21450      F   пенсионер     0        155999          сделка с автомобилем   
21451      M   сотрудник     1         89672                  недвижимость   
21452      M   сотрудник     1        244093  на покупку своего автомобиля   
21453      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21449                     B  операции с недвижимостью  
21450                     C    операции с автомобилем  
21451                     C  операции с недвижимостью  
21452                     B    операции с автомобилем  
21453                     C    операции с автомобилем  

[21454 rows x 12 columns]

Нулевых результатов нет.
Все цели считаны и категорированы

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Основные цели кредита определены правильно! 👍
</div>

### Ответы на вопросы.

##### Вопрос 1:

Есть ли зависимость между количеством детей и возвратом кредита в срок?

##### Вывод 1:

In [20]:
children_debt = df.pivot_table(index='children',values='debt',aggfunc=lambda a: np.mean(a) * 100) #Лямбда ф-ция для перевода в читаемый вид
display(children_debt)

debt
children          
0         7.543822
1         9.165808
2         9.492481
3         8.181818
4         9.756098
5         0.000000

Мы можем увидеть, что наиболее рискованными заемщиками являются те, у кого `2` или `4` ребенка, однако выборка клиентов с `4` детьми слишком мала, исходя из этого нельзя утверждать, что `4` ребенка имеют больше риска чем `2`.
Так как наименее рискованная категория, это заемщики без детей, что логично, так как ребенок не приносит дохода, а только требует вложений.

Возможно, то что люди с `2` детьми наиболее рискованные заемщики означает тот факт, что обычно в семьи с `2` и более детьми,находятся в более уязвимом положении, так как в случае болезни одного ребенка, это сразу распространяется и на второго, что приводит к дополнительным тратам.

Для людей с `5` детьми слишком маленькая выборка для оценки из за чего по данной группе людей отсутствуют просрочки.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Верно!
</div>

##### Вопрос 2: 

Есть ли зависимость между семейным положением и возвратом кредита в срок?

##### Вывод 2


In [21]:
family_debt = df.pivot_table(index='family_status_id',values='debt',aggfunc=lambda a: np.mean(a) * 100) #Лямбда ф-ция для перевода в читаемый вид
display(family_debt)

debt
family_status_id          
0                 7.545182
1                 9.347145
2                 6.569343
3                 7.112971
4                 9.750890

Мы можем увидеть, что наиболее рискованный заемщик не женат / не замужем. Это может быть связано с тем, что в случае возникновения финансовых проблем, он не может расчитывать на помощь другого человека, а так же более склонен к необдуманным решениями в связи с отсутствием в жизни постоянной иной точки зрения

А наименее рискованный Вдова/Вдовец, что достаточно очевидно, так как в случае смерти супруга, второй член семьи получает наследство и таким образом концентрирует у себя накопления семьи.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Согласен! Идём дальше)
</div>

##### Вопрос 3:

Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

##### Вывод 3:

In [22]:
income_debt = df.pivot_table(index='total_income_category',values='debt',aggfunc=lambda a: np.mean(a) * 100) #Лямбда ф-ция для перевода в читаемый вид
display(income_debt)

debt
total_income_category          
A                      8.000000
B                      7.062091
C                      8.491508
D                      6.000000
E                      9.090909

Мы можем увидеть, что наиболее рискованный заемщик с доходом от `0 до 30 000` рублей в месяц. Это заемщик с наиболее низким доходом из представленных категорий, для данного типа заемщиков любое финансовое потрясение может привести к просрочке, так же у данной категории нет накоплений. 

А наименее рискованный с доходом от `20 000 до 50 000` рублей в месяц. Обычно это люди имеющие стабильную работу, не склонные риску и понимающие факт того, что у них невысокий уровень дохода.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Тут всё правильно
</div>

##### Вопрос 4:

Как разные цели кредита влияют на его возврат в срок?

##### Вывод 4:

In [23]:
purpose_debt = df.loc[df['debt'] == 1,['purpose_category','debt']].groupby('purpose_category').sum() #Выделим посчитаем число задолженностей по цели кредита
total_purpose = df[['purpose_category','debt']].groupby('purpose_category').count() #Посчитаем общее число заемщиков по цели кредита
purpose_sum = total_purpose.merge(purpose_debt, on='purpose_category',how='left') #Объединим число просрочек и число заемщиков
purpose_sum['risk'] = (purpose_sum['debt_y'] / purpose_sum['debt_x']) * 100 #Добавим уровень риска в процентном отношении задолженности к общему количеству
purpose_sum.drop(['debt_x','debt_y'], inplace=True, axis=1) #Удалим число должников и число заемщиков
display(purpose_sum)

risk
purpose_category                  
операции с автомобилем    9.359034
операции с недвижимостью  7.233373
получение образования     9.220035
проведение свадьбы        8.003442

In [24]:
purpose_debt = df.loc[df['debt'] == 1,['purpose_category','debt']].groupby('purpose_category').sum() #Выделим посчитаем число задолженностей по цели кредита
total_purpose = df[['purpose_category','debt']].groupby('purpose_category').count() #Посчитаем общее число заемщиков по цели кредита
purpose_sum = total_purpose.merge(purpose_debt, on='purpose_category',how='left') #Объединим число просрочек и число заемщиков
purpose_sum['risk'] = (purpose_sum['debt_y'] / purpose_sum['debt_x']) * 100 #Добавим уровень риска в процентном отношении задолженности к общему количеству
purpose_sum.drop(['debt_x','debt_y'], inplace=True, axis=1) #Удалим число должников и число заемщиков
display(purpose_sum)

risk
purpose_category                  
операции с автомобилем    9.359034
операции с недвижимостью  7.233373
получение образования     9.220035
проведение свадьбы        8.003442

In [25]:
purpose_debt_v2 = df.pivot_table(index='purpose_category',values='debt',aggfunc=lambda a: np.mean(a) * 100) #Лямбда ф-ция для перевода в читаемый вид
display(purpose_debt_v2)

debt
purpose_category                  
операции с автомобилем    9.359034
операции с недвижимостью  7.233373
получение образования     9.220035
проведение свадьбы        8.003442

Мы можем увидеть, что наиболее рискованный заемщик с целью: `Операции с автомобилем`. Автокредит как таковой не только требует ежемесячной оплаты, но постоянного содержания, а так же как движимое имущество, обладает повышенным риском к повреждениям, что так же может выливаться в дополнительные траты.

Наименее рискованный заемщик - `Операции с недвижимостью`. Недвижимость - долгосрочный займ, обеспеченный этой недвижимостью. Обычно характеризуются невысоким ежемесячным платежом в связи с длинным сроком. Коммерческая ипотека так же направлена на то, что объект недвижимости становится доходом, что так же уменьшает риск невыплаты кредита.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b>
<br>Абсолютно верно!
</div>

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера v1:</b> 
<br>У тебя верные расчеты, но давай потренируемся строить сводные таблицы используя метод pivot_table. Построй её для ответа на любой из вопросов и в расчитай внутри таблицы долю должников внутри каждой группы. С этим может помочь aggfunc='mean'
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Так действительно намного проще, чем по сути составлять pivot_table вручную. 
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b> 
<br>Можно сделать прямо совсем коротко:

	df.pivot_table(index=['purpose_category'], values='debt', aggfunc='mean')
</div>

## Общий вывод:

Проанализировав данные, мы можем сделать следующие выводы касательно качества заемщиков.
**Наиболее рискованный заемщик в каждой группе:**

2 ребенка

Не женат или не замужем

С доходом менее 30 000 тысяч

Кредит связанный с автомобилем

**Наименее рискованный заемщик по каждой группе:**

Без детей

Вдовец/Вдова

С доходом от 30000 до 50000 

Под недвижимость

Из для более точной оценки желательно было бы получить полную стоимость кредита по каждому заемщику, так как без привязки к суммам, тяжело управлять этим риском. Однако это дает примерно представление о наиболее благополучной категории заемщиков и наименее благополучной категории. 
На основании исследования стоит внести тестовые коррективы в скоринговую модель и собрать данные по отказам и их причинам и сравнить с данными по отказам до тестирования новой скоринговой модели. 
Крайне желательно провести исследование доходности каждой группы клиентов и сравнить с риском, невозвратов, вполне возможна ситуация, когда тот же автокредит, хоть и является наиболее рисковым, будет приносить даже в рамках реализации риска суммарно больше прибыли, чем кредит на свадьбу.



<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v1:</b> 
<br>Общий вывод верный 👍
</div>

<div class="alert alert-danger">
<b>❌ Итоговый комментарий ревьюера v1:</b> 
<br>Ты хорошо потрудился, большая часть работы сделана, но остаётся внести следующие правки:
<br>- построить сводную таблицу используя метод pivot_table() для любого из вопросов в шаге 3. В таблице нужно произвести расчет доли должников внутри каждой группы клиентов. 
</div>

<div class="alert alert-success">
<b>✔️ Итоговый комментарий ревьюера v2:</b>
<br>Ты проделал большую работу! У тебя точные выводы с хорошей аргументацией. Здорово, что ты пишешь свои рассуждения и предположения - это помогает видеть ход твоей мысли. Отдельно хочу выделить аккуратность и структурированность твоей работы. Рад был поработать над проверкой твоего проекта! Желаю удачи в дальнейшем обучении! 😉
</div>